### ISSUE TITLE
version 1.5.0 install failure in win10+pyhon3.6.1

### ISSUE THREAD

**Issue by gdyxml2000** [tensorflow/tensorboard#916]
version 1.5.0 install failure in win10+pyhon3.6.1,version 0.4 is normal。
version 1.5.0 install error info:
Unknown requires Python '>=2.6, <3' but the running Python is 3.6.1

that  led me to install tensorflow1.5 also failed！！

**Comment by fortran-favors-the-old** [comment]
Expanding on https://github.com/tensorflow/tensorboard/issues/916#issue-292172148 with some more detail.  In [`setup.py`](https://github.com/tensorflow/tensorboard/blob/1.5.0/tensorboard/pip_package/setup.py#L36) it looks like `futures` shouldn't be required for Python < 3.2, but despite me using Python 3.6.3, it's being pulled in anyway.  The problem is that [`futures` is not compatible with Python 3](https://github.com/agronholm/pythonfutures/blob/3.2.0/setup.py#L33).

As https://github.com/tensorflow/tensorboard/issues/916#issue-292172148 points out, this breaks installation of `tensorflow`.

```
[fortran-favors-the-old@RHEL7-host] ~ % /opt/rh/rh-python36/root/usr/bin/python -m venv pyvenv
[fortran-favors-the-old@RHEL7-host] ~ % source pyvenv/bin/activate
(pyvenv) [fortran-favors-the-old@RHEL7-host] ~ % python --version
Python 3.6.3
(pyvenv) [fortran-favors-the-old@RHEL7-host] ~ % pip freeze --all
pip==9.0.1
setuptools==36.5.0
(pyvenv) [fortran-favors-the-old@RHEL7-host] ~ % pip install tensorflow-tensorboard==1.5.0
Collecting tensorflow-tensorboard==1.5.0
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/43/69/82e2a368076c94edbba3cd15804103bf1f31486d69e11551b71fa1d1f384/tensorflow_tensorboard-1.5.0-py3-none-any.whl (3.0MB)
    100% |████████████████████████████████| 3.0MB 336kB/s
Collecting six>=1.10.0 (from tensorflow-tensorboard==1.5.0)
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/67/4b/141a581104b1f6397bfa78ac9d43d8ad29a7ca43ea90a2d863fe3056e86a/six-1.11.0-py2.py3-none-any.whl
Collecting werkzeug>=0.11.10 (from tensorflow-tensorboard==1.5.0)
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/20/c4/12e3e56473e52375aa29c4764e70d1b8f3efa6682bef8d0aae04fe335243/Werkzeug-0.14.1-py2.py3-none-any.whl (322kB)
    100% |████████████████████████████████| 327kB 2.5MB/s
Collecting protobuf>=3.4.0 (from tensorflow-tensorboard==1.5.0)
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/40/99/471fa05dab1cf69419c91bcd7b5a7f3a0e251c76025bbdb40e08c367b728/protobuf-3.5.1-cp36-cp36m-manylinux1_x86_64.whl (6.4MB)
    100% |████████████████████████████████| 6.4MB 178kB/s
Collecting wheel>=0.26 (from tensorflow-tensorboard==1.5.0)
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/0c/80/16a85b47702a1f47a63c104c91abdd0a6704ee8ae3b4ce4afc49bc39f9d9/wheel-0.30.0-py2.py3-none-any.whl (49kB)
    100% |████████████████████████████████| 51kB 3.0MB/s
Collecting numpy>=1.12.0 (from tensorflow-tensorboard==1.5.0)
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/dc/ac/5c270dffb864f23315e9c1f9e0a0b300c797b3c170666c031c4de42aacae/numpy-1.14.0-cp36-cp36m-manylinux1_x86_64.whl (17.2MB)
    100% |████████████████████████████████| 17.2MB 68kB/s
Collecting futures>=3.1.1 (from tensorflow-tensorboard==1.5.0)
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/1f/9e/7b2ff7e965fc654592269f2906ade1c7d705f1bf25b7d469fa153f7d19eb/futures-3.2.0.tar.gz
futures requires Python '>=2.6, <3' but the running Python is 3.6.3
```

**Comment by fortran-favors-the-old** [comment]
For what it's worth, the issue still occurs even after ensuring that both `pip` and `setuptools` are each up to date:

```
(pyvenv) [fortran-favors-the-old@RHEL7-host] ~ % python --version
Python 3.6.3
(pyvenv) [fortran-favors-the-old@RHEL7-host] ~ % pip freeze --all
pip==9.0.1
setuptools==38.4.0
(pyvenv) [fortran-favors-the-old@RHEL7-host] ~ % pip install tensorflow-tensorboard==1.5.0
Collecting tensorflow-tensorboard==1.5.0
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/43/69/82e2a368076c94edbba3cd15804103bf1f31486d69e11551b71fa1d1f384/tensorflow_tensorboard-1.5.0-py3-none-any.whl (3.0MB)
    100% |████████████████████████████████| 3.0MB 336kB/s
Collecting wheel>=0.26 (from tensorflow-tensorboard==1.5.0)
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/0c/80/16a85b47702a1f47a63c104c91abdd0a6704ee8ae3b4ce4afc49bc39f9d9/wheel-0.30.0-py2.py3-none-any.whl (49kB)
    100% |████████████████████████████████| 51kB 3.1MB/s
Collecting html5lib==0.9999999 (from tensorflow-tensorboard==1.5.0)
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/ae/ae/bcb60402c60932b32dfaf19bb53870b29eda2cd17551ba5639219fb5ebf9/html5lib-0.9999999.tar.gz (889kB)
    100% |████████████████████████████████| 890kB 1.1MB/s
Collecting six>=1.10.0 (from tensorflow-tensorboard==1.5.0)
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/67/4b/141a581104b1f6397bfa78ac9d43d8ad29a7ca43ea90a2d863fe3056e86a/six-1.11.0-py2.py3-none-any.whl
Collecting futures>=3.1.1 (from tensorflow-tensorboard==1.5.0)
  Downloading https://internal-nexus-server.tld/repository/pypi-proxy/packages/1f/9e/7b2ff7e965fc654592269f2906ade1c7d705f1bf25b7d469fa153f7d19eb/futures-3.2.0.tar.gz
futures requires Python '>=2.6, <3' but the running Python is 3.6.3
```

**Comment by fortran-favors-the-old** [comment]
You can fix item 1 above (and hence resolve this issue) by updating `setuptools` to the latest version before running `setup.py`.  [`setuptools` 36.2.0 fixed a bug](http://setuptools.readthedocs.io/en/latest/history.html#v36-2-0) that's generating invalid requirements metadata in your wheel.

Here's an example illustrating the problem:

```
Example project

% cat > setup.py << 'EOF'
import setuptools

setuptools.setup(
    name='foo',
    version='1.0.0',
    install_requires=[
        'futures >= 3.1.1;python_version < "3.2"',
    ],
)
EOF


With broken setuptools

% rm -fr foo
% /opt/rh/rh-python36/root/usr/bin/python -m venv foo
% source foo/bin/activate
% python --version
% pip install --upgrade pip setuptools==36.1.1 wheel
% pip freeze --all
pip==9.0.1
setuptools==36.1.1
wheel==0.30.0
% rm -fr build dist foo-1.0.0.dist-info foo.egg-info
% python setup.py bdist_wheel
% deactivate
% unzip dist/foo-1.0.0-py3-none-any.whl
% grep -e '^Requires-Dist:' foo-1.0.0.dist-info/METADATA
Requires-Dist: futures (>=3.1.1)


With fixed setuptools

% rm -fr foo
% /opt/rh/rh-python36/root/usr/bin/python -m venv foo
% source foo/bin/activate
% python --version
% pip install --upgrade pip setuptools==36.2.0 wheel
% pip freeze --all
pip==9.0.1
setuptools==36.2.0
wheel==0.30.0
% rm -fr build dist foo-1.0.0.dist-info foo.egg-info
% python setup.py bdist_wheel
% deactivate
% unzip dist/foo-1.0.0-py3-none-any.whl
% grep -e '^Requires-Dist:' foo-1.0.0.dist-info/METADATA
Requires-Dist: futures (>=3.1.1); python_version < "3.2"
```

You'll also want to adjust the conditions for requiring `futures`, since `futures > 3.1.1` isn't compatible with Python 3 at all.

**Comment by nfelt** [comment]
Thank you for your in-depth report, I will take a look.

**Comment by jart** [comment]
Note that [TensorFlow Python support matrix](https://www.tensorflow.org/install/install_linux) is officially Python 2.7 and Python 3.4+. If TensorBoard worked on Python 3.[01] with the setup.py fix I'd be surprised.

**Comment by fortran-favors-the-old** [comment]
Great, in that case it sounds like fixing this will be more straightforward.

**Comment by nfelt** [comment]
@fortran-favors-the-old What is your pip version?  It looks like problem 2 that you identified is addressed in the most recent versions of pip (9.0.0+) by the PEP 503 support here:
https://github.com/pypa/pip/commit/c2101982cb5b28963fe22792cbb30ce50988ab8d

Namely, the newer pip should not try to download and install a version of a package known to be incompatible with the current python version if other versions of that package exist that are compatible.  On my machine with pip 9.0.1 this manifests itself as reliably choosing futures 3.1.1 over 3.2.0.

To be clear, I'm not disputing that we should still ensure this works for older pip versions and that issue 1 is a problem in its own right, but just confirming my understanding of the issue.

**Comment by nfelt** [comment]
Ah, I just saw in your original message that you appear to also be using pip 9.0.1.  In that case it's a bit strange you're seeing this - maybe your internal proxy could be affecting this behavior?

Could you try re-running the commands that you did in the "Here's what happens when I attempt to install tensorflow-tensorboard==1.5.0 on a RHEL 7 host with Python 3.6.3" section but use `pip install -vvv tensorflow-tensorboard==1.5.0` instead?

My pip shows the following message and I'm a bit surprised yours doesn't:
```
Skipping link https://pypi.python.org/packages/2d/99/b2c4e9d5a30f6471e410a146232b4118e697fa3ffc06d6a65efde84debd0/futures-3.2.0-py2-none-any.whl#md5=cfd62ab6c9852b04bb6048480fefaa75 (from https://pypi.python.org/simple/futures/) (requires-python:>=2.6, <3); it is not compatible with this Python
```

**Comment by fortran-favors-the-old** [comment]
Hey @nfelt , thanks for looking at this.

I won’t have access to my computer until the end of the day, so I’ll show you the output I get from running your suggested command later.

That said, I’m not surprised by your output above. That wheel has `Requires-Python` metadata correctly populated, which pip is picking up on in your case (and perhaps mine too; I’m guessing I would have seen that too if I had passed verbose flags like you).

I suspect that in your case, assuming `futures` installation eventually succeeds for you, that pip eventually tries to install the 3.1.1 sdist and succeeds, since 3.1.1 is compatible with Python 3.

In my case, for whatever reason, pip eventually attempts to install 3.2.0 sdist (since there’s no 3.2.0 Python 3 wheel to attempt first). Since it’s an sdist, pip can’t see any `Python-Requires` metadata ahead of time, so it attempts installation from the sdist, which eventually fails due to blacklisting Python 3 in setup.py . When package installation fails at this point, pip errors out.

When I observed this inconsistency in which package versions are attempted is when I really started really scratching my head when making my initial report. I probably should have mentioned back then what I’m about to say now, but I didn’t want to overload you with information.

[PEP 440]() has this IMO unfortunate line:

```
When multiple candidate versions match a version specifier, the preferred version SHOULD be the latest version as determined by the consistent ordering defined by the standard Version scheme.
```

**Comment by fortran-favors-the-old** [comment]
About 30 minutes ago I accidentally submitted my previous comment halfway though writing it. It’s since been updated. Sorry for any confusion that may have caused.

**Comment by fortran-favors-the-old** [comment]
@nfelt : Following up on the request you made yesterday:

When PyPI is the package index, `Python-Requires` metadata is available even for sdists (my initial guess about this was thankfully incorrect), and pip knows to skip the `futures` 3.2.0 sdist:

```
The package https://pypi.python.org/packages/1f/9e/7b2ff7e965fc654592269f2906ade1c7d705f1bf25b7d469fa153f7d19eb/futures-3.2.0.tar.gz#md5=d1b299a06b96ccb59f70324716dc0016 (from https://pypi.python.org/simple/futures/) (requires-python:>=2.6, <3) is incompatible with the pythonversion in use. Acceptable python versions are:>=2.6, <3
```

However, our internal Nexus mirror does not expose this metadata (bug!), so it isn't skipped.  pip then prefers 3.2.0 over 3.1.1 since it's the most recent version:

```
Using version 3.2.0 (newest of versions: 3.1.1, 3.2.0)
```

Thanks for submitting a pull request!

**Comment by pquentin** [comment]
Another data point for what it's worth: we have the same issue and we also use a PyPI mirror (devpi). Thanks @fortran-favors-the-old and @nfelt!

**Comment by fortran-favors-the-old** [comment]
@pquentin : FYI I noticed the same regarding devpi and plan on filing a bug report with them later today.

**Comment by nfelt** [comment]
Okay, good to know we were able to track down that discrepancy as well - thanks for following up!

The PR to fix the futures dependency to be python2-only has been submitted, and so it should go out in a 1.5.1 bugfix release fairly soon (I anticipate probably no later than end-of-week).

**Comment by fortran-favors-the-old** [comment]
Great, thanks for the quick fix @nfelt !

**Comment by nfelt** [comment]
Closing this out, since the fix should now be in the 1.5.1 release: https://github.com/tensorflow/tensorboard/releases/tag/1.5.1
